Source/Core/Core/WC24PatchEngine.cpp

```diff
@@ -0,0 +1,136 @@
+    Titles::NINTENDO_CHANNEL_NTSC_J,
+    Titles::NINTENDO_CHANNEL_PAL,
+    Titles::FORECAST_CHANNEL_NTSC_U,
+    Titles::FORECAST_CHANNEL_NTSC_J,
+    Titles::FORECAST_CHANNEL_PAL,
+    Titles::NEWS_CHANNEL_NTSC_U,
+    Titles::NEWS_CHANNEL_NTSC_J,
+    Titles::NEWS_CHANNEL_PAL,
+    Titles::EVERYBODY_VOTES_CHANNEL_NTSC_U,
+    Titles::EVERYBODY_VOTES_CHANNEL_NTSC_J,
+    Titles::EVERYBODY_VOTES_CHANNEL_PAL,
+};
+
+static std::vector<NetworkPatch> s_patches;
+
+bool DeserializeLine(const std::string& line, NetworkPatch* patch)
+{
+  const std::vector<std::string> items = SplitString(line, ':');
+
+  if (items.size() != 3)
+    return false;
+
+  patch->source = items[0];
+  patch->replacement = items[1];
+
+  if (!TryParse(items[2], &patch->is_kd))
+    return false;
+
+  return patch;
+}
+
+void LoadPatchSection(const Common::IniFile& ini)
+{
+  std::vector<std::string> lines;
+  NetworkPatch patch;
+  ini.GetLines("WC24Patch", &lines);
+
+  for (std::string& line : lines)
+  {
+    if (line.empty())
+      continue;
+
+    if (line[0] == '$')
+    {
+      patch.name = line.substr(1, line.size() - 1);
+    }
+    else
+    {
+      if (DeserializeLine(line, &patch))
+      {
+        s_patches.push_back(patch);
+      }
+    }
+  }
+
+  ReadEnabledAndDisabled(ini, "WC24Patch", &s_patches);
+}
+
+void LoadPatches()
+{
+  const auto& sconfig = SConfig::GetInstance();
+  // We can only load WC24 Channels.
+  if (!IsWC24Channel())
+    return;
+
+  Common::IniFile ini;
+  // If WiiLink is enabled then we load the Default Ini as that has the WiiLink URLs.
+  if (Config::Get(Config::MAIN_WII_WIILINK_ENABLE))
+    ini = sconfig.LoadDefaultGameIni();
+  else
+    ini = sconfig.LoadLocalGameIni();
+
+  LoadPatchSection(ini);
+}
+
+bool IsWC24Channel()
+{
+  const auto& sconfig = SConfig::GetInstance();
+  const auto found =
+      std::find(s_wc24_channels.begin(), s_wc24_channels.end(), sconfig.GetTitleID());
+
+  return found != s_wc24_channels.end();
+}
+
+void Reload()
+{
+  s_patches.clear();
+  LoadPatches();
+}
+
+std::optional<std::string> GetNetworkPatch(const std::string& source, IsKD is_kd)
+{
+  const auto patch =
+      std::find_if(s_patches.begin(), s_patches.end(), [&source, &is_kd](NetworkPatch& patch) {
+        return patch.source == source && patch.is_kd == is_kd;
+      });
+  if (patch == s_patches.end())
+    return std::nullopt;
+
+  return patch->replacement;
+}
+
+// When we patch for the Socket, we aren't given the URL. Instead, it is in a Host header
+// in the payload that the socket is going to send. We need to manually find which patch to apply.
+std::optional<std::string> GetNetworkPatchByPayload(std::string_view source)
+{
+  for (const NetworkPatch& patch : s_patches)
+  {
+    if (source.find(patch.source) != std::string::npos && patch.is_kd != IsKD{true})
+      // We found the correct patch, return it!
+      return ReplaceAll(std::string{source}, patch.source, patch.replacement);
```

Comment:
This also seems too aggressive. You'd be replacing any instances outside the Host header, too.

Reply:
This on the other hand is still clearly wrong. You're targetting the `Host:` part of the HTTP header here, yeah? https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages In that case, you should also only search in that area of the request. Basically, something like (untested if this actually works, you might want to write a unit test for the logic because it seems prone to off-by-one errors...)

```
  size_t pos = 0;
  while (pos < source.size())
  {
    if (source[pos] == '\n')  // end of header
      break;

    if (source.substr(pos).starts_with("Host:"))
    {
      size_t end_of_line = source.find('\n', pos);
      std::string_view domain =
          source.substr(pos + 5, end_of_line == std::string_view::npos ? std::string_view::npos :
                                                                         (end_of_line - pos - 5));
      for (const NetworkPatch& patch : s_patches)
      {
        if (domain.trim() == patch.source)
        {
          return source.substr(0, pos) + "Host: " + patch.replacement +
                 (end_of_line == std::string_view::npos ? "" : source.substr(end_of_line));
        }
      }
    }
    else
    {
      size_t end_of_line = source.find('\n', pos);
      if (end_of_line == std::string_view::npos)
        break;
      pos = end_of_line + 1;
    }
  }
  return std::nullopt;
```

Reply:
This example was a big help. With some testing and looking at [RFC2616](https://www.rfc-editor.org/rfc/rfc2616#section-2.2), all HTTP messages have CRLF line endings, so I opted for checking for that instead of just `\n`. Other than that and the return line, it works perfectly with my testing in the Nintendo Channel.